spark/src/main/java/org/apache/iceberg/spark/OrcSchemaWithTypeVisitorSpark.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.iceberg.MetadataColumns;
+import org.apache.iceberg.orc.ORCSchemaUtil;
+import org.apache.iceberg.orc.OrcSchemaWithTypeVisitor;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Types;
+import org.apache.orc.TypeDescription;
+
+public abstract class OrcSchemaWithTypeVisitorSpark<T> extends OrcSchemaWithTypeVisitor<T> {
```

Comment:
why do we need a new visitor that only modifies the struct visitor (as opposed to modify the parent/existing one)? Are there cases where we would still want to use the parent visitor (i.e., not handling default values)?

Reply:
My rationale is that, the parent is a more generic ORC visitor, whereas the ones I intend to modify is only related to spark behavior (cause the behavior is dependent on/relevant to `buildOrcProjection`. I only want the overridden behavior to affect on the scope of spark, that's why I structure my code this way.

Reply:
I see. The goal, however, is make Iceberg support default values for AVRO and ORC formats, not only for Spark. The changes for AVRO are along these lines.

Reply:
@rzhang10 , is the logic implemented in a way that only supports Spark behavior?

Reply:
Actually, it is not just for spark, it supports the default value for ORC regardless of the engine, I just don't feel like changing the original visitor code to minimize the potential impact, but we could as well just modify the `OrcSchemaWithTypeVisitor` code itself without using an extended subclass.

Reply:
if the support for default values is not limited by the engine type, then it is fine from that perspective to use a subclass. However, for code consistency, we need to use the same methodology for both union types and default values, for ORC  (i.e.,  either both modify or both extend the visitor). I will leave it to you @rzhang10  and @funcheetah  to decide which way makes more sense).

Reply:
Actually, after my new patch (which uses the `BaseDataReader.convertConstant` to convert iceberg constant to spark constant), the visitor has become Spark-specific, since its visit method directly call `convertConstant` which is spark specific conversion logic. So this can adds to the reason why I stick with my current class inheritance structure.

Reply:
that is fine for now. Can you please create an issue to support default values for ORC, for non-spark readers?